File: pkg/virt/host.go
/*
* This file is part of Celestial (https://github.com/OpenFogStack/celestial).
* Copyright (c) 2024 Tobias Pfandzelter, The OpenFogStack Team.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, version 3.
*
* This program is distributed in the hope that it will be useful, but
* WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
* General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program. If not, see <http://www.gnu.org/licenses/>.
**/

package virt

import (
	"io/fs"
	"os"
	"os/exec"

	"github.com/pkg/errors"
	log "github.com/sirupsen/logrus"
)

// initHost resets the hosts iptables and sets up basics on the host.
// Partially based on https://github.com/firecracker-microvm/firecracker-demo/blob/main/one-time-setup.sh
func (v *Virt) initHost() error {

	// clear iptables
	// iptables -F
	cmd := exec.Command(IPTABLES_BIN, "-F")

	if out, err := cmd.CombinedOutput(); err != nil {
		return errors.Wrapf(err, "%#v: output: %s", cmd.Args, out)
	}

	// sh -c "echo 1 > /proc/sys/net/ipv4/ip_forward"

	file, err := os.Create("/proc/sys/net/ipv4/ip_forward")

	if err != nil {
		return err
	}

	defer func(file *os.File) {
		err := file.Close()
		if err != nil {
			log.Error(err.Error())
		}
	}(file)

	if _, err := file.WriteString("1"); err != nil {
		return errors.WithStack(err)
	}

	// Configure packet forwarding
	// sysctl -wq net.ipv4.conf.all.forwarding=1

	cmd = exec.Command(SYSCTL_BIN, "-wq", "net.ipv4.conf.all.forwarding=1")

	if out, err := cmd.CombinedOutput(); err != nil {
		return errors.Wrapf(err, "%#v: output: %s", cmd.Args, out)
	}

	// Avoid "neighbour: arp_cache: neighbor table overflow!"
	// sysctl -wq net.ipv4.neigh.default.gc_thresh1=1024
	cmd = exec.Command(SYSCTL_BIN, "-wq", "net.ipv4.neigh.default.gc_thresh1=1024")

	if out, err := cmd.CombinedOutput(); err != nil {
		return errors.Wrapf(err, "%#v: output: %s", cmd.Args, out)
	}

	// sysctl -wq net.ipv4.neigh.default.gc_thresh2=2048
	cmd = exec.Command(SYSCTL_BIN, "-wq", "net.ipv4.neigh.default.gc_thresh2=2048")

	if out, err := cmd.CombinedOutput(); err != nil {
		return errors.Wrapf(err, "%#v: output: %s", cmd.Args, out)
	}

	// sysctl -wq net.ipv4.neigh.default.gc_thresh3=4096
	cmd = exec.Command(SYSCTL_BIN, "-wq", "net.ipv4.neigh.default.gc_thresh3=4096")

	if out, err := cmd.CombinedOutput(); err != nil {
		return errors.Wrapf(err, "%#v: output: %s", cmd.Args, out)
	}

	//sudo iptables -t nat -A POSTROUTING -o [HOSTINTERFACE] -j MASQUERADE

	cmd = exec.Command(IPTABLES_BIN, "-t", "nat", "-A", "POSTROUTING", "-o", v.hostInterface, "-j", "MASQUERADE")

	if out, err := cmd.CombinedOutput(); err != nil {
		return errors.Wrapf(err, "%#v: output: %s", cmd.Args, out)
	}

	// sudo iptables -A FORWARD -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT

	cmd = exec.Command(IPTABLES_BIN, "-A", "FORWARD", "-m", "conntrack", "--ctstate", "RELATED,ESTABLISHED", "-j", "ACCEPT")

	if out, err := cmd.CombinedOutput(); err != nil {
		return errors.Wrapf(err, "%#v: output: %s", cmd.Args, out)
	}

	// prepare an output folder

	// prepare the file writers for output
	// if the directory exists, remove it
	if _, err = os.Stat(OUTPUTPATH); err == nil {
		err = os.RemoveAll(OUTPUTPATH)

		if err != nil {
			return err
		}
	}

	// then create it
	err = os.Mkdir(OUTPUTPATH, fs.FileMode(0755))
	if err != nil {
		return err
	}

	// finally, check that the clocksource is tsc and set it otherwise
	clocksource, err := os.ReadFile("/sys/devices/system/clocksource/clocksource0/current_clocksource")

	if err != nil {
		return err
	}

	if string(clocksource) != "tsc\n" {
		// set the clocksource to tsc
		log.Warnf("The current clock source on the host is %s, setting it to tsc", string(clocksource))
		err = os.WriteFile("/sys/devices/system/clocksource/clocksource0/current_clocksource", []byte("tsc"), fs.FileMode(0644))

		if err != nil {
			return err
		}
	}

	return nil
}

File: celestial/host.py
#
# This file is part of Celestial (https://github.com/OpenFogStack/celestial).
# Copyright (c) 2024 Tobias Pfandzelter, The OpenFogStack Team.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#

"""Adapter for communication with the Celestial hosts over gRPC"""

import typing
import grpc
import time
import logging

import proto.celestial.celestial_pb2
import proto.celestial.celestial_pb2_grpc


class Host:
    """
    Communication link for a Celestial host.
    """

    def __init__(self, num: int, addr: str):
        """
        Initialize host communication.

        :param num: The host number.
        :param addr: The address of the host.
        """
        self.num = num
        self.addr = addr

        c = grpc.insecure_channel(self.addr)
        self.stub = proto.celestial.celestial_pb2_grpc.CelestialStub(c)

        self.public_key = ""

    def register(self) -> proto.celestial.celestial_pb2.RegisterResponse:
        """
        Send a `register` request to the host.

        :return: The response from the host.
        """
        try:
            request = proto.celestial.celestial_pb2.RegisterRequest(host=self.num)

            response: proto.celestial.celestial_pb2.RegisterResponse = (
                self.stub.Register(request)
            )

        except Exception as e:
            logging.error(f"Error registering host {self.num}: {e}")
            exit(1)

        # others currently not used
        self.peer_public_key = response.peer_public_key

        self.peer_listen_addr = (
            self.addr.split(":")[0] + ":" + response.peer_listen_addr.split(":")[1]
        )

        logging.debug(f"host {self.num} registered")
        logging.debug(f"memory: {response.available_ram}")
        logging.debug(f"cpu: {response.available_cpus}")

        return response

    def init(
        self,
        init_request: proto.celestial.celestial_pb2.InitRequest,
    ) -> None:
        """
        Send an `init` request to the host.

        :param hosts: A list of all hosts in the constellation.
        :param machines: A dictionary mapping host numbers to a list of machine ID and machine configuration tuples.
        """

        self.stub.Init(init_request)

        return

    def stop(self) -> None:
        """
        Send a `stop` request to the host.
        """
        self.stub.Stop(proto.celestial.celestial_pb2.Empty())

    def update(
        self,
        update_requests: typing.Iterator[
            proto.celestial.celestial_pb2.StateUpdateRequest
        ],
    ) -> None:
        """
        Send a `update` request to the host.

        :param machine_diff: An iterator of machine ID and machine state tuples.
        :param link_diff: An iterator of link tuples.
        """

        t1 = time.perf_counter()
        self.stub.Update(update_requests)
        t2 = time.perf_counter()
        logging.debug(f"update transmission took {t2-t1} seconds")

        return
